### ISSUE TITLE
[RFC][DistMuon] Handle fused weights correctly for Kimi K3 per-head Muon

### ISSUE THREAD

**Issue by JavaZeroo** [pytorch/torchtitan#4692]
## Problem

Muon orthogonalizes one logical matrix at a time. Several K3 parameters fuse more than one logical matrix into a single tensor, and DistMuon has to see the same matrix boundaries the forward pass uses. On upstream/main it does not, and K3 has no per-head Muon recipe at all (`kimi_k3_debugmodel` still trains with AdamW; #4272 P0.5).

## Fused projections in K3

Only the parameters Muon would own. Norms, embeddings, `lm_head`, KDA `beta` / `A_log` / `dt_bias` / convs and the vision tower stay on AdamW.

| parameter | logical weights and how they are stored | what NS must see | after upstream changes |
|---|---|---|---|
| `feed_forward.w13`, `moe.shared_experts.w13` | `w1 [F, D]` and `w3 [F, D]`, stored as `[2F, D]` with rows interleaved `w1_0, w3_0, w1_1, w3_1, ...` (#4535) | 2 matrices `[F, D]` | `[2, F, D]` stacked, FSDP shards dim 1 (#4676): native batch of 2 |
| `moe.routed_experts.*.w1_EFD`, `w3_EFD` | `w1_e [F, D]` and `w3_e [F, D]` per expert, stored as two `[E, F, D]` tensors | 2E matrices `[F, D]` | `w13 [E, 2, F, D]` (#4539, #4643): native batch of 2E |
| `attention.wkv_b` | `K_nope_h [nope, C]` and `V_h [v, C]` per head, stored as `[H * (nope + v), C]` with rows `K_nope_0, V_0, K_nope_1, V_1, ...` (each block contiguous) | 2H matrices: H of `[nope, C]`, H of `[v, C]` | unchanged |
| `attention.wkv_a` | `W_kv_latent [latent, D]` and `W_k_rope [rope, D]`, stored as `[latent + rope, D]`, latent rows then rope rows | 2 matrices `[latent, D]`, `[rope, D]` | unchanged |
| `attention.wq_b` | `Q_nope_h [nope, R]` and `Q_rope_h [rope, R]` per head, stored as `[H * (nope + rope), R]` with rows `Q_nope_0, Q_rope_0, Q_nope_1, ...` | 2H matrices: H of `[nope, R]`, H of `[rope, R]` | unchanged |
| `delta_attention.q_proj`, `k_proj`, `v_proj`, `forget_b`, `output_gate` | one `[d, *]` matrix per head, stored as `[H * d, *]` | H matrices `[d, *]` | unchanged |

K3 sizes: debugmodel `H=16, nope=64, rope=32, v=64, latent=256, d=128`; Kimi-K3 `H=96, nope=128, rope=64, v=128, latent=512, d=128`.

Kimi runs one Newton-Schulz per logical projection for MLA, so `wq_b`, `wkv_a` and `wkv_b` are split as above and each part gets its own aspect-ratio lr adjustment (Su Jianlin, https://kexue.fm/archives/11126/comment-page-2#comment-28601).

## What we build on

- #4676 turns `w13` into a stacked `[2, F, D]` tensor. #4539 / #4643 do the same for routed experts (`[E, 2, F, D]`) and make DistMuon accept native `[..., R, C]` batches. Once both land, interleaving is gone from the FFN path. DistMuon on upstream/main already distributes a batch-first parameter with `Shard(0)` compute (one complete matrix per rank, the same path routed experts use), so `w13` needs no extra layout metadata; I checked a `[2, F, D]` parameter stored `Shard(1)` against per-matrix `torch.optim.Muon` and the update matches exactly.
- #4531 fixes the NS shape for fused gated linears and, after its rebase on #4676, drops `MatrixBatchLayout` in favor of the native batch. Nothing below depends on it.

## Gap: `torch.split` with unequal sizes

What DistMuon on upstream/main can express, written as the view it takes before Newton-Schulz:

```python
# BlockShard(block_size=R): equal blocks, one matrix per head
w.view(-1, R, C)                          # [M, R, C]
# native batch (w13 after #4676, experts after #4643): leading batch dims
w13                                       # [2, F, D] or [E, 2, F, D]
```

What `wq_b`, `wkv_b` and `wkv_a` need:

```python
# wq_b [H * (nope + rope), R]
q_nope, q_rope = torch.split(wq_b.view(H, nope + rope, R), [nope, rope], dim=1)  # [H, nope, R], [H, rope, R]
# wkv_b [H * (nope + v), C]
k_nope, v = torch.split(wkv_b.view(H, nope + v, C), [nope, v], dim=1)            # [H, nope, C], [H, v, C]
# wkv_a [latent + rope, D]
kv_latent, k_rope = torch.split(wkv_a, [latent, rope], dim=0)                    # [latent, D], [rope, D]
```

Every existing layout is a fixed-size `view`; there is no way to say `split([nope, v])`. `BlockShard(block_size=nope)` only works while the two parts happen to be equal (`wkv_b` in K3), and nothing works for `nope + rope` or `latent + rope`.

Proposal: let `BlockShard.block_size` be a tuple, `BlockShard(dim=0, block_size=(nope, v))`: one block of `nope + v` rows made of these contiguous matrices. `BlockShard` keeps transporting whole heads, so redistribution planning is unchanged; at compute time DistMuon does the `split` above through narrowed views, runs NS once per piece, and applies the aspect-ratio lr adjustment per piece since the shapes differ. `wkv_a` is one block `(latent, rope)`. It only touches `BlockShard`, so it does not depend on #4531 or on the `w13` layout. Implemented in #4693 with a 2-GPU test, plus a single-GPU comparison against `torch.optim.Muon` on the split matrices (both K3 head dims and an unequal variant).

## Plan

0. Recipe-helper prerequisites split out of #4596 at review request, both on the K2.7 recipe and independent of everything else: `muon_lr` / `adamw_lr` passed separately (#4755) and the bucket construction simplified (#4756).
1. `[DistMuon] Split BlockShard blocks into matrix segments` (#4693) -- independent PR on upstream/main. Its second commit switches the K2.7 recipe to the split layouts, since K2.7 has the same gap today. Verified on 2x H100: the 2-GPU segment test passes, K2.5 FSDP2+EP2 runs, same-commit runs are bitwise deterministic, main vs #4693 differs by at most 1.9e-2 in loss over 60 steps (step 1 identical), FSDP2 vs FSDP2+EP2 by at most 2.4e-3.
2. `[kimi k3] Switch K3 recipe to per-head DistMuon` (#4596) -- branch is main + #4693 + the recipe commit; lands after #4676 (`w13` becomes a `[2, F, D]` batch, `Shard(0)` compute) and #4643 (routed experts become `w13 [E, 2, F, D]`). Per-head layouts for MLA and KDA, `wq_b` / `wkv_b` / `wkv_a` through tuple block sizes. DistMuon rejects TP, so the B200 TP test keeps AdamW and a new `kimi_k3_mm_muon` test (fsdp2 x ep2) covers the default optimizer. Adds a CPU test that every matrix parameter is either matched by the Muon pattern with a compute layout or is on an explicit AdamW list, so a rename can no longer drop a weight to AdamW silently. Loss comparison on B200 (KDA needs SM100).

cc @tianyu-l @weifengpy @shuhuayu @WhoisZihan @anijain2305 @sanketpurandare @rakkit

**Comment by rakkit** [comment]
thx.
table looks reasonable, i think FFN/MoE wouldn't  be the problem.

for attention, wq_b, wkv_a and wkv_b, each weights gonna split into 2 (H) matrixes, you wanna sent ", wq_b, wkv_a, wkv_b" to 3 ranks or the split 2(H) matrixs to 6 (or even more) ranks, or make this flexible?  cause its a jointly scaling of model depth and fsdp x pp degree.   you may not have enough of ranks to run only one matrix's NS5 per rank(which also does not really make sense, batched NS5 is pretty efficient).

**Comment by JavaZeroo** [comment]
> thx. table looks reasonable, i think FFN/MoE wouldn't be the problem.
> 
> for attention, wq_b, wkv_a and wkv_b, each weights gonna split into 2 (H) matrixes, you wanna sent ", wq_b, wkv_a, wkv_b" to 3 ranks or the split 2(H) matrixs to 6 (or even more) ranks, or make this flexible? cause its a jointly scaling of model depth and fsdp x pp degree. you may not have enough of ranks to run only one matrix's NS5 per rank(which also does not really make sense, batched NS5 is pretty efficient).

It's batched, not one matrix per rank. BlockShard distributes whole blocks over the dp_shard ranks, so with `H=96` and `8 ranks` each rank gets 12 heads of wq_b and 12 of wkv_b. Inside a rank the block batch is split into its pieces and NS runs once per piece as a batched baddbmm: one `[12, nope, C]` batch and one `[12, v, C]` batch.

**Comment by shuhuayu** [comment]
related comments saying k2 splitting these logical matrices: https://kexue.fm/archives/11126/comment-page-2
